### PR TITLE
Feat/stripe controller AB#114183

### PR DESCRIPTION
**Pull request by hamsaaldrobi**
AB#114183

### Type:
* [ ] cicd: helm, docker & cicd adjustments.
* [x] feature: new capabilities.
* [ ] fix: bug, hotfix, etc.
* [ ] refactor: enhacements.
* [ ] style: changes in styles.
* [ ] other: docs, tests.

### What's the focus of this PR:
- Process: 
- 1. We get a payload from `http2sb` into this service
- 2. We handle the payload according if its paid or not
- 3. If yes, we get its transactions
- 4. If has transactions, we publish the new **enriched payload** as the original payload + transactions

### CODE REVIEW

File: components/controller/controller.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        logger.info(`messageId ${messageId} | Fetched ${transactions ? transactions.length : 0} transactions for the payout ${payoutId}`);
+
+        if (transactions) {
+          const chargedTransactions = getChargedTransactions(transactions);
+          logger.info(`messageId ${messageId} | Found ${chargedTransactions ? chargedTransactions.length : 0} charge transactions | shop ${shop} | payoutId ${payoutId}`);
+        }
+
+        const payoutTransformed = { ...payout, transactions };
+        await publishPayoutTransformed(payoutTransformed);
+        trackMetric('payout_transformed');
```
Comment (Betisman) [inline components/controller/controller.js:46]:
```suggestion
      let chargedTransactions = [];
      try {
        const transactions = await payment.getBalanceTransactionsByPayout(shop, payoutId) || [];
        logger.info(`messageId ${messageId} | Fetched ${transactions ? transactions.length : 0} transactions for the payout ${payoutId}`);

       chargedTransactions = getChargedTransactions(transactions);
        logger.info(`messageId ${messageId} | Found ${chargedTransactions ? chargedTransactions.length : 0} charge transactions | shop ${shop} | payoutId ${payoutId}`);

        const payoutTransformed = { ...payout, chargedTransactions };
        await publishPayoutTransformed(payoutTransformed);
        trackMetric('payout_transformed');
```

You're filtering the transactions but ignoring the result.

File: test/components/controller/controller.test.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+          await controller.handlePayout('noordhoff')(incomingMessage);
+        } catch (err) {
+          error = err;
+        } finally {
+          expect(error).toBeUndefined();
+          expect(metrics.trackMetric).toHaveBeenCalledWith(expect.objectContaining({ name: 'payout_transformed' }));
+        }
+      });
+    });
+    describe('from Plantyn', () => {
```
Comment (Betisman) [inline test/components/controller/controller.test.js:78]:
```suggestion

    describe('from Plantyn', () => {
```

File: components/controller/controller.js (hunk 1 above)
Comment (hamsaaldrobi) [reply components/controller/controller.js:46]: so should i publish the charged transactions instead of the transactions itself?

**Comment by Betisman** [comment]
I miss some tests to test that the transformation (adding the transactions to the payout) is working fine.
And a test to test that the request to Stripe to get the transactions fails.

File: components/controller/controller.js (hunk 1 above)
Comment (Betisman) [reply components/controller/controller.js:46]: If you don't publish the charged transactions, why filter the full array?

File: components/controller/controller.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    const publishPayoutTransformed = initPublication('payoutTransformed');
+    const publishPayoutFailedToTransform = initPublication('payoutFailedToTransform');
+
+    const trackMetric = event => metrics.trackMetric({ name: event, value: 1 });
+
+    const getChargedTransactions = transactions => transactions.filter(({ source }) => source.object === 'charge');
+
+    const handlePayout = shop => async message => {
+      const { messageId } = message.properties;
+      const { body: payout } = message;
```
Comment (neodmy) [inline components/controller/controller.js:16]:
```suggestion
      const { properties: { messageId }, body: payout } = message;
```

Comment (hamsaaldrobi) [reply components/controller/controller.js]: Nice catch 👁️

File: components/controller/controller.js (hunk 1 above)
Comment (Betisman) [reply components/controller/controller.js:45]: If we decide to publish just the raw transactions, then we don't need that filter.

Comment (Betisman) [reply components/controller/controller.js:45]:
```suggestion
      try {
        const transactions = await payment.getBalanceTransactionsByPayout(shop, payoutId);
        logger.info(`messageId ${messageId} | Fetched ${transactions ? transactions.length : 0} transactions for the payout ${payoutId}`);

        const payoutTransformed = { ...payout, transactions };
        await publishPayoutTransformed(payoutTransformed);
        trackMetric('payout_transformed');
```

File: package.json [hunk 4]
```diff
@@ -35,22 +35,29 @@
     "supertest": "^6.1.3"
   },
   "dependencies": {
+    "@azure/data-tables": "^1.0.0-beta.4",
```
Comment (neodmy) [inline package.json:38]: Do we need this dependency?

File: components/controller/controller.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        const transactions = await payment.getBalanceTransactionsByPayout(shop, payoutId);
+        logger.info(`messageId ${messageId} | Fetched ${transactions ? transactions.length : 0} transactions for the payout ${payoutId}`);
+
+        if (transactions) {
+          const chargedTransactions = getChargedTransactions(transactions);
+          logger.info(`messageId ${messageId} | Found ${chargedTransactions ? chargedTransactions.length : 0} charge transactions | shop ${shop} | payoutId ${payoutId}`);
+        }
+
+        const payoutTransformed = { ...payout, transactions };
+        await publishPayoutTransformed(payoutTransformed);
```
Comment (neodmy) [inline components/controller/controller.js:44]: I'd add the compression. Payouts can potentially be big

```suggestion
        await publishPayoutTransformed(payoutTransformed, { contentEncoding: 'zlib' });
```

File: package.json (hunk 4 above)
Comment (hamsaaldrobi) [reply package.json:38]: nope at all, fixing it

File: components/controller/controller.js (hunk 5 above)
Comment (Betisman) [reply components/controller/controller.js:44]: +1

File: test/components/controller/controller.test.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
+/* eslint-disable no-empty */
+/* eslint-disable jest/no-disabled-tests */
```
[the remark this answers is not in the document]
Comment (hamsaaldrobi) [reply test/components/controller/controller.test.js]: done

File: components/controller/controller.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+      }
+
+      try {
+        const transactions = await payment.getBalanceTransactionsByPayout(shop, payoutId);
+        logger.info(`messageId ${messageId} | Fetched ${transactions ? transactions.length : 0} transactions for the payout ${payoutId}`);
+
+        const createdISODate = new Date(createUnixTimestamp * 1000).toISOString();
+        const updatedISODate = new Date(updateUnixTimestamp * 1000).toISOString();
+
+        const payoutTransformed = {
```
Comment (Betisman) [inline components/controller/controller.js:41]: This is ok for now, but I would refactor this snippet to create a proper transformation component.

File: components/controller/controller.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        trackMetric('payout_skipped');
+        return;
+      }
+
+      try {
+        const transactions = await payment.getBalanceTransactionsByPayout(shop, payoutId);
+        logger.info(`messageId ${messageId} | Fetched ${transactions ? transactions.length : 0} transactions for the payout ${payoutId}`);
+
+        const createdISODate = new Date(createUnixTimestamp * 1000).toISOString();
+        const updatedISODate = new Date(updateUnixTimestamp * 1000).toISOString();
```
Comment (Betisman) [inline components/controller/controller.js:39]: I guess you can refactor these and create a function so you can reuse it instead of duplicating code. :)

Comment (hamsaaldrobi) [reply components/controller/controller.js:39]: Okay in progress 👍🏼

File: components/controller/controller.js (hunk 7 above)
Comment (hamsaaldrobi) [reply components/controller/controller.js:41]: Yeah i got you, i'll fix this as well setting a transformation component

**Pull request merged**